File: tests/OrderedSetTest.cpp
// Copyright 2021 Ian Ruh
#include <gtest/gtest.h>
#include <symengine/basic.h>
#include <symengine/symbol.h>

#include "OrderedSet.h"

using SymEngine::Basic;
using SymEngine::RCP;
using SymEngine::Symbol;

TEST(OrderedSetTests, Basics) {
    RCP<const Symbol> x = SymEngine::symbol("x");
    RCP<const Symbol> y = SymEngine::symbol("y");
    RCP<const Symbol> z = SymEngine::symbol("z");

    cppmpc::OrderedSet set;

    EXPECT_TRUE(set.isConsistent());
    set.append(x);
    EXPECT_TRUE(set.isConsistent());
    set.append(y);
    EXPECT_TRUE(set.isConsistent());
    set.append(z);

    EXPECT_EQ(set.at(0), x);
    EXPECT_EQ(set.at(1), y);
    EXPECT_EQ(set.at(2), z);

    EXPECT_EQ(3, set.size());

    set.remove(0);
    EXPECT_TRUE(set.isConsistent());
    EXPECT_EQ(set.at(0), y);
    EXPECT_EQ(set.at(1), z);

    EXPECT_EQ(2, set.size());
}

TEST(OrderedSetTests, Union) {
    RCP<const Symbol> x = SymEngine::symbol("x");
    RCP<const Symbol> y = SymEngine::symbol("y");
    RCP<const Symbol> z = SymEngine::symbol("z");

    cppmpc::OrderedSet set1;
    cppmpc::OrderedSet set2;

    set1.append(x);
    set1.append(y);

    set2.append(y);
    set2.append(z);

    EXPECT_TRUE(set1.isConsistent());
    EXPECT_TRUE(set2.isConsistent());

    set1.unionWith(set2);

    EXPECT_TRUE(set1.isConsistent());
    EXPECT_EQ(3, set1.size());
}

TEST(OrderedSetTests, Subset) {
    RCP<const Symbol> x = SymEngine::symbol("x");
    RCP<const Symbol> y = SymEngine::symbol("y");
    RCP<const Symbol> z = SymEngine::symbol("z");

    cppmpc::OrderedSet set1;
    cppmpc::OrderedSet set2;
    cppmpc::OrderedSet set3;

    set1.append(x);
    set1.append(y);

    set2.append(y);
    set2.append(z);

    set3.append(x);

    EXPECT_TRUE(set1.isSubset(set3));
    EXPECT_FALSE(set3.isSubset(set2));
    EXPECT_FALSE(set1.isSubset(set2));
}

File: tests/ordered_set_test.py
import unittest
import symengine
import cppmpc


class OrderedSetTests(unittest.TestCase):
    def test_basic(self):
        x = symengine.Symbol("x")
        y = symengine.Symbol("y")
        z = symengine.Symbol("z")

        test_set = cppmpc.OrderedSet()

        test_set.append(x)
        self.assertTrue(test_set.isConsistent())
        test_set.append(y)
        self.assertTrue(test_set.isConsistent())
        test_set.append(z)
        self.assertTrue(test_set.isConsistent())

        self.assertEqual(x, test_set.at(0))
        self.assertEqual(y, test_set.at(1))
        self.assertEqual(z, test_set.at(2))

        self.assertEqual(test_set.size(), 3)

        test_set.remove(0)
        self.assertTrue(test_set.isConsistent())

        self.assertEqual(y, test_set.at(0))
        self.assertEqual(z, test_set.at(1))

    def test_union(self):
        x = symengine.Symbol("x")
        y = symengine.Symbol("y")
        z = symengine.Symbol("z")

        test_set1 = cppmpc.OrderedSet()
        test_set2 = cppmpc.OrderedSet()

        test_set1.append(x)
        test_set1.append(y)

        test_set1.append(y)
        test_set2.append(z)

        self.assertTrue(test_set1.isConsistent())
        self.assertTrue(test_set2.isConsistent())

        test_set1.unionWith(test_set2)

        self.assertTrue(test_set1.isConsistent())
        self.assertEqual(test_set1.size(), 3)

    def test_subset(self):
        x = symengine.Symbol("x")
        y = symengine.Symbol("y")
        z = symengine.Symbol("z")

        test_set1 = cppmpc.OrderedSet()
        test_set2 = cppmpc.OrderedSet()
        test_set3 = cppmpc.OrderedSet()

        test_set1.append(x)
        test_set1.append(y)

        test_set1.append(y)
        test_set2.append(z)

        test_set3.append(x)

        self.assertTrue(test_set1.isSubset(test_set3))
        self.assertFalse(test_set3.isSubset(test_set2))
        self.assertFalse(test_set1.isSubset(test_set2))


if __name__ == "__main__":
    unittest.main()
